storage/memory.js

```diff
@@ -0,0 +1,349 @@
+'use strict'
+
+const joi = require('joi')
+const LRUCache = require('mnemonist/lru-cache')
+const nullLogger = require('abstract-logging')
+const StorageInterface = require('./interface')
+const { findMatchingIndexes, findNotMatching, bsearchIndex } = require('../util')
+
+const DEFAULT_CACHE_SIZE = 1024
+
+const optionsValidation = joi.object({
+  size: joi.number().integer().min(1).default(DEFAULT_CACHE_SIZE),
+  log: joi.object().default(() => nullLogger),
+  invalidation: joi.boolean().default(false)
+})
+
+/**
+ * @typedef StorageMemoryOptions
+ * @property {?number} [size=1024]
+ * @property {?Logger} [log]
+ * @property {?boolean} [invalidation=false]
+ */
+
+class StorageMemory extends StorageInterface {
+  /**
+   * in-memory storage
+   * @param {StorageMemoryOptions} options
+   */
+  constructor (opts) {
+    const { value: options, error } = optionsValidation.validate(opts || {})
+    if (error) { throw error }
+
+    super(options)
+    this.size = options.size
+    this.log = options.log
+    this.invalidation = options.invalidation
+
+    this.init()
+  }
+
+  init () {
+    this.store = new LRUCache(this.size)
+
+    if (!this.invalidation) {
+      return
+    }
+    // key -> references, keys are strings, references are sorted array strings
+    this.keysReferences = new Map()
+    // same as above, but inverted
+    this.referencesKeys = new Map()
+  }
+
+  /**
+   * retrieve the value by key
+   * @param {string} key
+   * @returns {undefined|*} undefined if key not found or expired
+   */
+  async get (key) {
```

Comment:
```suggestion
 get (key) {
```

Reply:
the functions on `storage` are all async because of redis io, shouldn't they be consistent?